cel/pkg/reconciler/celrun/celrun.go

```diff
@@ -0,0 +1,283 @@
+	if run.Spec.Ref != nil && run.Spec.Ref.Name != "" {
+		// Use the k8s client to get the CEL rather than the lister.  This avoids a timing issue where
+		// the CEL is not yet in the lister cache if it is created at nearly the same time as the Run.
+		// See https://github.com/tektoncd/pipeline/issues/2740 for discussion on this issue.
+		cs, err := c.CELClientSet.CustomV1alpha1().CELs(run.Namespace).Get(ctx, run.Spec.Ref.Name, metav1.GetOptions{})
+		if err != nil {
+			run.Status.MarkRunFailed(celv1alpha1.CELRunReasonCouldntGetCEL.String(),
+				"Error retrieving CEL for Run %s/%s: %s",
+				run.Namespace, run.Name, err)
+			return nil, nil, fmt.Errorf("error retrieving CEL for Run %s: %w", fmt.Sprintf("%s/%s", run.Namespace, run.Name), err)
+		}
+		CELMeta = cs.ObjectMeta
+		CELSpec = cs.Spec
+	} else {
+		// Run does not require name but for CEL it does
+		run.Status.MarkRunFailed(celv1alpha1.CELRunReasonCouldntGetCEL.String(),
+			"Missing spec.ref.name for Run %s/%s",
+			run.Namespace, run.Name)
+		return nil, nil, fmt.Errorf("missing spec.ref.name for Run %s", fmt.Sprintf("%s/%s", run.Namespace, run.Name))
+	}
+	return &CELMeta, &CELSpec, nil
+}
+
+func storeCELSpec(status *celv1alpha1.CELStatus, cs *celv1alpha1.CELSpec) {
+	// Only store the CELSpec once, if it has never been set before
+	if status.Spec == nil {
+		status.Spec = cs
+	}
+}
+
+func propagateCELLabelsAndAnnotations(run *v1alpha1.Run, CELMeta *metav1.ObjectMeta) {
+	// Propagate labels from CEL to Run
+	if run.ObjectMeta.Labels == nil {
+		run.ObjectMeta.Labels = make(map[string]string, len(CELMeta.Labels)+1)
+	}
+	for key, value := range CELMeta.Labels {
+		run.ObjectMeta.Labels[key] = value
+	}
+	run.ObjectMeta.Labels[cel.GroupName+CELLabelKey] = CELMeta.Name
+	run.ObjectMeta.Labels[KindLabelKey] = cel.ControllerName
+
+	// Propagate annotations from CEL to Run
+	if run.ObjectMeta.Annotations == nil {
+		run.ObjectMeta.Annotations = make(map[string]string, len(CELMeta.Annotations))
+	}
+	for key, value := range CELMeta.Annotations {
+		run.ObjectMeta.Annotations[key] = value
+	}
+}
+
+func (r *Reconciler) updateLabelsAndAnnotations(ctx context.Context, run *v1alpha1.Run) error {
+	newRun, err := r.runLister.Runs(run.Namespace).Get(run.Name)
+	if err != nil {
+		return fmt.Errorf("error getting Run %s when updating labels/annotations: %w", run.Name, err)
+	}
+	if !reflect.DeepEqual(run.ObjectMeta.Labels, newRun.ObjectMeta.Labels) || !reflect.DeepEqual(run.ObjectMeta.Annotations, newRun.ObjectMeta.Annotations) {
+		mergePatch := map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"labels":      run.ObjectMeta.Labels,
+				"annotations": run.ObjectMeta.Annotations,
+			},
+		}
+		patch, err := json.Marshal(mergePatch)
+		if err != nil {
+			return err
+		}
+		_, err = r.pipelineClientSet.TektonV1alpha1().Runs(run.Namespace).Patch(ctx, run.Name, types.MergePatchType, patch, metav1.PatchOptions{})
+		return err
+	}
+	return nil
```

Comment:
readability: Similar vein to https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow, typically it's recommended to front-load early exit conditions where you can, instead of wrapping them in an if.

e.g. 

```go
r := Get()
if reflect.DeepEqual(labels) && reflect.DeepEqual(annotations) {
  // no change, nothing else to do
  return nil
}
// continue on with patch
```